Kokkos Kernels: adding missing TPLs and pre-conditions

lucbv opened the pull request:
Adding variants and dependencies for rocBLAS and rocSPARSE. Also adding a "when=" close to the TPL variants that prevents enabling the TPLs in versions of the library when it was not yet available.

lucbv committed:
Kokkos Kernels: adding missing TPLs and pre-conditions

Adding variants and dependencies for rocBLAS and rocSPARSE.
Also adding a "when=" close to the TPL variants that prevents
enabling the TPLs in versions of the library when it was not
yet available.

lucbv committed:
Kokkos Kernels: remove comment for better format

lucbv committed:
Kokkos Kernels: fix issue with unpacking wrong number of args

After changing the tpls dictionary we need to update the unpacking
logic to catch the right number of outputs out of it!

cedricchevalier19 reviewed (commented):
Thanks for the information.
It's definitely a nice improvement.

cedricchevalier19 commented:
I let @brian-kelley approving as with Kokkos we are not consumers of Kokkos-Kernels.

lucbv commented:
Thanks @cedricchevalier19 and @brian-kelley let us wait for a Spack maintainer to have a look so we can merge this.
This will set us up a bit better to generate our modules on Frontier and Livermore's machines : )

lucbv commented:
@alecbcs @tldahlgren could you have quick look at this please, it is ready to be merged

tldahlgren commented on var/spack/repos/builtin/packages/kokkos-kernels/package.py:
```diff
@@ -136,19 +136,21 @@ class KokkosKernels(CMakePackage, CudaPackage):
 
     tpls = {
         # variant name   #deflt   #spack name   #root var name #docstring
```
Do you want to add the condition to this line so it matches the change?

lucbv replied:
Yes, updated the comment to add the condition up there

tldahlgren commented on var/spack/repos/builtin/packages/kokkos-kernels/package.py:
```diff
@@ -136,19 +136,21 @@ class KokkosKernels(CMakePackage, CudaPackage):
-        deflt_bool, spackname, rootname, descr = tpls[tpl]
-        variant(tpl, default=deflt_bool, description=descr)
+        deflt_bool, spackname, rootname, condition, descr = tpls[tpl]
+        variant(tpl, default=deflt_bool, when="%s" % condition, description=descr)
```
FYI. You *could* use a python f-string here (see https://docs.python.org/3/tutorial/inputoutput.html#formatted-string-literals).  We're converting the core to use them as we make changes.

lucbv replied:
Sure f-string is a bit simpler and more readable so switched to that. Assuming this all goes well I might create a later PR that goes through the CMake logic section of the package and switch it to f-string as well.

tldahlgren reviewed (dismissed):
LGTM though I highly recommend adding the condition to the comments describing the contents of the tuple.

tldahlgren commented:
Let me know if you want to merge as-is though I do recommend updating the comment first.

lucbv committed:
Kokkos Kernels: updating doc string for tpls var and using f-string

Improving comment a bit and switching to f-string for more readability.
When setup to do more testing will try to use f-string in the CMake
options generation part of the package.

lucbv committed:
Style change

tldahlgren reviewed (approved):
Thanks.
